File: daily/rust/15-3-sum.rs
fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
    let mut nums = nums;
    nums.sort();
    let mut res = vec![];
    for i in 0..nums.len() {
        if i > 0 && nums[i] == nums[i - 1] {
            continue;
        }
        let (mut l, mut r) = (i + 1, nums.len() - 1);
        while l < r {
            let sum = nums[i] + nums[l] + nums[r];
            if sum < 0 {
                l += 1;
            } else if sum > 0 {
                r -= 1;
            } else {
                res.push(vec![nums[i], nums[l], nums[r]]);
                while l < r && nums[l] == nums[l + 1] {
                    l += 1;
                }
                while l < r && nums[r] == nums[r - 1] {
                    r -= 1;
                }
                l += 1;
                r -= 1;
            }
        }
    }
    res
}

#[test]
fn test1() {
    let nums = vec![-1, 0, 1, 2, -1, -4];
    let res = three_sum(nums);
    assert_eq!(res, vec![vec![-1, -1, 2], vec![-1, 0, 1]]);
}

#[test]
fn test2() {
    let nums = vec![];
    let res: Vec<Vec<i32>> = three_sum(nums);
    assert_eq!(res, vec![]);
}

File: daily/cpp/15-3-sum.cpp
#include <bits/stdc++.h>

#include <vector>
using namespace std;

class Solution {
 public:
  vector<vector<int>> threeSum(vector<int>& nums) {
    vector<vector<int>> res;
    int n = nums.size(), l, r, sum;
    sort(nums.begin(), nums.end());
    for (int i = 0; i < nums.size(); ++i) {
      if (i > 0 && nums[i - 1] == nums[i]) continue;

      l = i + 1;
      r = n - 1;

      while (l < r) {
        sum = nums[i] + nums[l] + nums[r];
        if (sum < 0) {
          l++;
        } else if (sum > 0) {
          r--;
        } else {
          res.push_back({nums[i], nums[l], nums[r]});
          while (l < r && nums[l] == nums[l + 1]) l++;
          while (r > l && nums[r] == nums[r - 1]) r--;
          l++;
          r--;
        }
      }
    }
    return res;
  }
};